omod/src/main/java/org/openmrs/web/controller/maintenance/SearchIndexController.java

```diff
@@ -48,12 +51,33 @@ public String showPage() {
 		Map<String, Object> results = new HashMap<String, Object>();
 		log.debug("rebuilding search index");
 		try {
-			Context.updateSearchIndex();
+			updateSearchIndexAsync = Context.updateSearchIndexAsync();
 		} catch (RuntimeException e) {
 			success = false;
 		}
 
 		results.put("success", success);
 		return results;
 	}
+
+	/**
+	 * @should return return inProgress for status if a rebuildSearchIndex is not completed
+	 * @should return success for status if a rebuildSearchIndex is completed successfully
+	 * @should return error for status if a rebuildSearchIndex is not completed normally
+	 * @return hashMap of String, String holds a key named "status" indicating the status of
+	 * rebuild search index
+	 */
+    @RequestMapping(method = RequestMethod.GET, value = "admin/maintenance/rebuildSearchIndexStatus")
+    public @ResponseBody Map<String, String> getStatus() {
+    	if(updateSearchIndexAsync == null)
+    		throw new APIException("There was a problem rebuilding the search index");
+
+        Map<String, String> results = new HashMap<>();
+        if (updateSearchIndexAsync.isDone()) {
+			results.put("status", updateSearchIndexAsync.isCancelled() ? "error" : "success");
+            updateSearchIndexAsync = null;
+        } else
```

Comment:
This code would be clearer and safe if you used {} braces even for the else clause as advised here https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-CodeStyle